pkg/kubelet/images/image_gc_manager.go

```diff
@@ -113,22 +113,23 @@ type imageCache struct {
 	images []container.Image
 }
 
-// set updates image cache.
+// set updates image cache using the given image list. imageCache
+// takes ownership of the list, you should not reference the list again
+// after calling this function.
 func (i *imageCache) set(images []container.Image) {
 	i.Lock()
 	defer i.Unlock()
 	i.images = images
 }
 
 // get gets a sorted (by image size) image list from image cache.
-// There is a potentical data race in this function. See PR #60448
-// Because there is deepcopy function available currently, move sort
-// function inside this function
 func (i *imageCache) get() []container.Image {
 	i.Lock()
 	defer i.Unlock()
-	sort.Sort(sliceutils.ByImageSize(i.images))
-	return i.images
+	// Make a copy of image list to avoid data race
+	images := append([]container.Image{}, i.images...)
```

Comment:
How frequently is this read? Should we sort on set and return the list without allocating?

Reply:
That's a good point! It's invoked in different places so it's hard to tell from source. I searched [this](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/kubelet_node_status.go#L294-L297) and [this](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/kubelet_node_status.go#L342-L363) log messages in a master node and found the [read](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/kubelet_node_status.go#L369) is invoked every few (<5) seconds (see the log below). Compared to the write getting invoked [every 30s](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/images/image_gc_manager.go#L187-L196), I agree it makes sense to sort on set. 

Regarding whether we should skip allocating on return, I'm not sure if it's safe to do. Currently we [truncate the length of the returned list](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/nodestatus/setters.go#L445), which has no effect on the original list https://play.golang.org/p/fc6JmArJtfW. But I still feel it's safer if we copy on return, to prevent the original list getting mutated in future. 

```
I0309 21:40:27.927125    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:35.971794    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:39.926950    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:43.582650    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:44.926879    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:45.926867    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:46.104704    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:52.927557    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:53.927034    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:56.239125    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:56.926903    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:06.404693    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:08.926829    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:11.926992    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:14.926864    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:16.541316    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:19.929088    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:20.926684    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:23.926947    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:23.927801    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:26.720531    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:28.926814    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:32.926843    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:32.927004    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:34.926880    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:36.884823    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:37.926909    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:38.926790    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:39.927096    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:43.582911    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:43.927033    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:47.013040    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:54.926907    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:55.927053    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:56.927574    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:57.161981    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:02.927006    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:04.926781    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:05.927033    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:07.335855    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:13.927023    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:17.502770    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:19.927029    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:21.926896    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:27.638850    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:32.926875    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
```